File: backend/src/main/java/com/sgenne/timetracking/project/service/ProjectService.java
package com.sgenne.timetracking.project.service;

import com.sgenne.timetracking.project.model.Project;
import com.sgenne.timetracking.project.repository.ProjectRepository;
import com.sgenne.timetracking.project.request.CreateProjectRequest;
import com.sgenne.timetracking.user.UserService;
import com.sgenne.timetracking.user.model.User;
import lombok.AllArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import javax.transaction.Transactional;

import static com.sgenne.timetracking.project.validation.ProjectValidator.descriptionIsValid;
import static com.sgenne.timetracking.project.validation.ProjectValidator.titleIsValid;
import static org.springframework.http.HttpStatus.BAD_REQUEST;

@Service
@AllArgsConstructor
@Transactional
public class ProjectService {

    private final ProjectRepository projectRepository;
    private final UserService userService;

    /**
     * Creates and stores a Project.
     * @param request Contains the information that will be used when creating the project.
     * @return The created project.
     */
    public Project createProject(CreateProjectRequest request) {
        String title = request.getTitle();
        String description = request.getDescription();
        Long ownerId = request.getOwnerId();
        User owner = userService.getUserById(ownerId);

        titleIsValid(title).orThrow(
                (message) -> new ResponseStatusException(BAD_REQUEST, message));

        descriptionIsValid(description).orThrow(
                (message) -> new ResponseStatusException(BAD_REQUEST, message));

        Project project = new Project(title, description, owner.getId());

        return projectRepository.save(project);
    }

    /**
     * Finds a stored Project.
     * @param projectId The id of the project to be found.
     * @return The found project.
     */
    public Project getProjectById(Long projectId) {
        return this
                .projectRepository
                .findById(projectId)
                .orElseThrow(
                        () -> new ResponseStatusException(
                                HttpStatus.NOT_FOUND,
                                String.format(
                                        "No project with project-id \"%s\" was found.",
                                        projectId)));
    }


}

File: backend/src/project/project.service.ts
import { randomUUID } from "crypto";
import { Repository } from "../db/repository";
import { Status } from "../service-result";
import ServiceResult from "../service-result/serviceResult.interface";
import { UserService } from "../user/user.service";
import Project from "./project.interface";

export interface ProjectService {
  createProject: (
    title: string,
    owner: string,
    description?: string
  ) => Promise<ServiceResult<Project>>;
  getProjectById: (id: string) => Promise<ServiceResult<Project>>;
}

export default class ProjectServiceProvider implements ProjectService {
  private readonly projectRepository: Repository<Project>;
  private readonly userService: UserService;

  constructor(
    projectRepository: Repository<Project>,
    userService: UserService
  ) {
    this.projectRepository = projectRepository;
    this.userService = userService;
  }

  /**
   * Creates a project.
   *
   * @param title The title of the new project.
   * @param ownerId The id of the owner of the new project.
   * @param description The optional description of the project.
   */
  async createProject(
    title: string,
    ownerId: string,
    description?: string
  ): Promise<ServiceResult<Project>> {
    const { payload: owner, status } = await this.userService.getUser(ownerId);

    if (!owner) {
      return {
        status: status,
        message: "No user with id ownerId was found.",
      };
    }

    const id = randomUUID();

    const project: Project = {
      title: title,
      ownerId: ownerId,
      description: description,
      projectId: id,
    };

    await this.projectRepository.create(project);

    return {
      status: Status.OK,
      payload: project,
    };
  }

  /**
   * Returns a specified project.
   *
   * @param id The id of the project to return.
   */
  async getProjectById(projectId: string): Promise<ServiceResult<Project>> {
    const foundProject: Project | undefined =
      await this.projectRepository.findOne({
        projectId: projectId,
      });

    if (!foundProject) {
      return {
        status: Status.RESOURCE_NOT_FOUND,
      };
    }

    return {
      status: Status.OK,
      payload: foundProject,
    };
  }
}
